2_process/src/calculate_toha_continuous.R

```diff
@@ -0,0 +1,232 @@
+  
+  # Divide by number of timesteps to get average area per day
+  oha_daily <- sum(oha_df$habitat)/nrow(oha_df) 
+  tha_daily <- tha_df$habitat/nrow(tha_df) # should be the same before/after division since tha_df is already daily
+  toha_daily <- sum(toha)/length(toha)
+  
+  return(data.frame(opti_hab = oha_daily, 
+                    therm_hab = tha_daily, 
+                    opti_therm_hab = toha_daily))
+  
+}
+
+# Vectorized function
+optical_habitat_area <- function(I_0, Kd, hypso, I_lower, I_upper) {
+  
+  # Expecting only one Kd per day but if not, needs to match length of I_0
+  stopifnot(length(Kd) == 1 || length(Kd) == length(I_0))
+  
+  z_max <- tail(hypso$depths, 1)
+  z_surface <- head(hypso$depths, 1)
+  
+  ##### Find exact depths of irr thresholds 
+  
+  Z1 <- log( I_upper / I_0) / -Kd
+  Z2 <- log( I_lower / I_0) / -Kd
+  
+  ##### Checks before calculating areas
+  completely_below_lake <- which(Z1 > z_max) # if OHA is completely below lake, benthic area is 0 (too bright)
+  completely_above_lake <- which(I_0 < I_lower) # if OHA is completely above lake, benthic area is 0 (too dark)
+  benth_0 <- c(completely_below_lake, completely_above_lake)
+  
+  # Set Zs to NA so area calc is quicker (?)
+  Z1[benth_0] <- NA 
+  Z2[benth_0] <- NA
+  
+  # if OHA extends above lake, use the surface as Z1 to calc OHA
+  extends_above_lake <- which(Z1 < z_surface)
+  Z1[extends_above_lake] <- z_surface
+  
+  # if OHA extends below lake, use the bottom as Z2 to calc OHA
+  extends_below_lake <- which(Z2 >= z_max) # also include those with Z2 == z_max
+  Z2[extends_below_lake] <- z_max
+  
+  ##### Create hypso at these different depths
+  A1 <- resample_hypso(hypso, Z1)[["areas"]]
+  A2 <- resample_hypso(hypso, Z2)[["areas"]]
+  
+  ##### Use areas and depths at thresholds to calculate benthic area
+  
+  benthic_area <- calc_benthic_area(Z1, Z2, A1, A2)
+  
+  ##### Final benthic area checks
+  
+  # Set benthic areas to zero based on above criteria
+  benthic_area[completely_below_lake] <- 0 
+  benthic_area[completely_above_lake] <- 0
+  
+  # Add lake bottom if bottom of optical habitat 
+  benthic_area[extends_below_lake] <- benthic_area[extends_below_lake] + tail(hypso$areas, 1)
+  
+  return(data.frame(Z1 = Z1, Z2 = Z2, habitat = benthic_area))
+}
+
+# Now vectorized (except for one apply fxn)
+thermal_habitat_area <- function(wtr_df, hypso, wtr_lower, wtr_upper) {
+  
+  z_max <- tail(hypso$depths, 1)
+  z_surface <- head(hypso$depths, 1)
+  z_wtr <- rLakeAnalyzer::get.offsets(wtr_df)
+  
+  wtr_surface <- wtr_df[[which(z_wtr == z_surface)]]
+  wtr_bottom <- wtr_df[[which(z_wtr == z_max)]]
+  
+  ##### Find exact depths of wtr thresholds 
+  Z1 <- apply(wtr_df, 1, function(wtr_row) {
+    # Using apply since these two actions need to be done per row
+    wtr <- as.vector(t(wtr_row))
+    approx(wtr, z_wtr, xout=wtr_upper)$y
+  })
+  
+  Z2 <- apply(wtr_df, 1, function(wtr_row) {
```

Comment:
Is it faster to do these together like this?
```r
Z1_Z2 <- apply(wtr_df, MARGIN = 1, function(wtr_row) {
    # Using apply since these two actions need to be done per row
    wtr <- as.vector(t(wtr_row))
    approx(wtr, z_wtr, xout=c(wtr_lower, wtr_upper))$y
})
```

Reply:
👍 just would need to flip wtr_lower and wtr_upper to make sure the correct things went to Z1 and Z2